sdk/core/src/retry_policy.rs

```diff
@@ -0,0 +1,123 @@
+use tokio::time::{sleep, Duration};
+
+#[async_trait::async_trait]
+pub trait RetryPolicy {
+    async fn create_instance(&self) -> Box<dyn RetryPolicyInstance>;
+}
+
+#[async_trait::async_trait]
+pub trait RetryPolicyInstance {
+    async fn retry(&mut self) -> bool;
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub struct LinearRetryPolicy {
+    max_retries: u32,
+    seconds_backoff: u32,
+}
+
+impl Default for LinearRetryPolicy {
+    fn default() -> Self {
+        Self {
+            max_retries: 3,
+            seconds_backoff: 2,
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl RetryPolicy for LinearRetryPolicy {
+    async fn create_instance(&self) -> Box<dyn RetryPolicyInstance> {
+        Box::new(LinearRetryPolicyInstance {
+            linear_retry_policy: self.clone(),
+            current_retries: 0,
+        })
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+struct LinearRetryPolicyInstance {
+    linear_retry_policy: LinearRetryPolicy,
+    current_retries: u32,
+}
+
+#[async_trait::async_trait]
+impl RetryPolicyInstance for LinearRetryPolicyInstance {
+    async fn retry(&mut self) -> bool {
+        self.current_retries += 1;
+        if self.current_retries <= self.linear_retry_policy.max_retries {
+            sleep(Duration::from_secs(
+                self.linear_retry_policy.seconds_backoff as u64,
+            ))
+            .await;
+            true
+        } else {
+            false
+        }
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub struct ExponentialBackoffRetryPolicy {
```

Comment:
I don't see why this is a separate policy. We have a single RetryPolicy across other SDKs that uses service-provided back-offs if supported, or just uses linear. Making individual clients decide or even customers just shifts the problem/responsibility down to developers when we can handle it upstream in core.

Reply:
I see. Good point. 
I'll remove the ExponentialBackoff and just leave linear and no retry policy.

Reply:
> and no retry policy

Not sure I follow. We should have retries as a policy as mentioned above. It makes customizing the policies in the pipeline straight forward with a consistent debugging pattern.

Reply:
mmm now I'm confused. In the [docs](https://docs.microsoft.com/en-us/azure/architecture/best-practices/retry-service-specific#retry-mechanism-10) we say:

> The built-in classes provide support for __linear__ (constant delay) and __exponential__ with randomization retry intervals. There is also a __no retry__ policy for use when another process is handling retries at a higher level. However, you can implement your own retry classes if you have specific requirements not provided by the built-in classes.

So I thought to code the very same policies here. 

Do you mean that we should have a single, customizable policy instead of 3 different policies?

Reply:
> Making individual clients decide or even customers just shifts the problem/responsibility down to developers when we can handle it upstream in core.

At least for C#, this is definitely the case: [https://docs.microsoft.com/en-us/dotnet/api/microsoft.azure.cosmos.table.tablerequestoptions.retrypolicy?view=azure-dotnet#Microsoft_Azure_Cosmos_Table_TableRequestOptions_RetryPolicy](https://docs.microsoft.com/en-us/dotnet/api/microsoft.azure.cosmos.table.tablerequestoptions.retrypolicy?view=azure-dotnet#Microsoft_Azure_Cosmos_Table_TableRequestOptions_RetryPolicy). 
Developers can choose which retry policy they want (they are provided a default but that's it).

Also, looking at the implementations, they have the aforementioned retry policies: [https://docs.microsoft.com/en-us/dotnet/api/microsoft.azure.cosmos.table.iretrypolicy?view=azure-dotnet](https://docs.microsoft.com/en-us/dotnet/api/microsoft.azure.cosmos.table.iretrypolicy?view=azure-dotnet):

* Microsoft.Azure.Cosmos.Table.ExponentialRetry
* Microsoft.Azure.Cosmos.Table.IExtendedRetryPolicy
* Microsoft.Azure.Cosmos.Table.LinearRetry
* Microsoft.Azure.Cosmos.Table.NoRetry

Arren't we supposed to copy this behaviour?

Reply:
Those are track 1 SDKs. For the most part, they are deprecated. Anything Microsoft.Azure.* is not worth looking at for guidance. Azure.* is. Our track 2 `RetryPolicy` is here: https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/core/Azure.Core/src/Pipeline/Internal/RetryPolicy.cs

You'll also notice where *it* does the retry loop, not the pipeline. This would be consistent with all our languages, where request goes through the policy chain and responses come out, but the pipeline doesn't interact with it directly. You could, for example, even take retries out without having to have a `NoRetryPolicy`.